src/System.Data.OleDb/src/System/Data/ProviderBase/DbConnectionFactory.cs

```diff
@@ -0,0 +1,591 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Data.ProviderBase
+{
+    internal abstract class DbConnectionFactory {
+        private Dictionary<DbConnectionPoolKey, DbConnectionPoolGroup>  _connectionPoolGroups;
+        private readonly List<DbConnectionPool> _poolsToRelease;
+        private readonly List<DbConnectionPoolGroup> _poolGroupsToRelease;
+        private readonly DbConnectionPoolCounters _performanceCounters;
+        private readonly Timer _pruningTimer;
+
+        private const int PruningDueTime =4*60*1000;           // 4 minutes
+        private const int PruningPeriod  =  30*1000;           // thirty seconds
+
+        private static int _objectTypeCount; // Bid counter
+        internal readonly int _objectID = System.Threading.Interlocked.Increment(ref _objectTypeCount);
+
+        // s_pendingOpenNonPooled is an array of tasks used to throttle creation of non-pooled connections to 
+        // a maximum of Environment.ProcessorCount at a time.
+        static int s_pendingOpenNonPooledNext = 0;
+        static Task<DbConnectionInternal>[] s_pendingOpenNonPooled = new Task<DbConnectionInternal>[Environment.ProcessorCount];
+        static Task<DbConnectionInternal> s_completedTask;
+
+        protected DbConnectionFactory() : this (DbConnectionPoolCountersNoCounters.SingletonInstance) { }
+
+        protected DbConnectionFactory(DbConnectionPoolCounters performanceCounters) {
+            _performanceCounters = performanceCounters;
+            _connectionPoolGroups = new Dictionary<DbConnectionPoolKey, DbConnectionPoolGroup>();
+            _poolsToRelease = new List<DbConnectionPool>();
+            _poolGroupsToRelease = new List<DbConnectionPoolGroup>();
+            _pruningTimer = CreatePruningTimer();
+        }
+
+        internal DbConnectionPoolCounters PerformanceCounters {
+            get { return _performanceCounters; }
+        }
+
+        abstract public DbProviderFactory ProviderFactory {
+            get;
+        }
+
+        internal int ObjectID {
+            get {
+                return _objectID;
+            }
+        }
+
+        public void ClearAllPools() {
+            // IntPtr hscp;
+            // Bid.ScopeEnter(out hscp, "<prov.DbConnectionFactory.ClearAllPools|API> ");
```

Comment:
All this `Bid.ScopeXXX` stuff can be removed, and the finally statements they required. And the `IntPtr`. I think it's fine to leave that til later if you prefer.

Again, thought I'd done this.

Reply:
Thinking about it, maybe just remove the `Bid.XXX` and `IntPtr hscp;` everywhere since it's a trivial search and remove the comments. The useless try/finally you could leave, as it will probably just JIT away to nothing, and if it doesn't no biggie someone can clean it up later if they want.